compiler/rustc_middle/src/ty/context.rs

```diff
@@ -3205,6 +3205,58 @@ impl<'tcx> TyCtxt<'tcx> {
         }
     }
 
+    fn get_expr_equivalent(self, expr: &hir::Expr<'_>) -> Option<DefId> {
```

Comment:
what is this actually trying to achieve? it would be best to spell it out here with a lot of comments because i'm not actually totally convinced that this needs to be implemented with a query.

Reply:
Like, are you trying to get the constructor def id of the `#[default]` variant for an ADT?

Reply:
@estebank: You responded in a comment but it's kinda difficult to have a contained convo over top-level comment threads so I'm going to respond here.

> [...] We keep track of the DefId of Vec::new. [...]

It seems a bit problematic to me that this only works for things that have a single ctor. But if we're working in that restriction model, then this should not be too difficult to derive from the MIR of the `default` impl which is already cross-crate encoded. Any valid "equivalence" should either have a single `_0 = {const}` assignment or `_0 = {call}` call terminator with no arguments from which we can very easily extract the def id.